OMCompiler/SimulationRuntime/c/simulation/solver/delay.c

```diff
@@ -185,27 +244,85 @@ double delayImpl(DATA* data, threadData_t *threadData, int exprNumber, double ex
       time1 = ((TIME_AND_VALUE*)getRingData(delayStruct, i+1))->t;
       value1 = ((TIME_AND_VALUE*)getRingData(delayStruct, i+1))->value;
     }
-    /* was it an exact match?*/
-    if(time0 == timeStamp){
-      infoStreamPrint(LOG_EVENTS_V, 0, "delayImpl: Exact match at %g = %g", timeStamp, value0);
-
+    /* Return left value */
+    if(time0 == timeStamp)
+    {
       return value0;
-    } else if(time1 == timeStamp) {
-      infoStreamPrint(LOG_EVENTS_V, 0, "delayImpl: Exact match at %g = %g", timeStamp, value1);
-
+    }
+    /* Return right value */
+    else if(time1 == timeStamp)
+    {
       return value1;
-    } else {
-      /* linear interpolation */
-      double timedif = time1 - time0;
-      double dt0 = time1 - timeStamp;
-      double dt1 = timeStamp - time0;
-      double retVal = (value0 * dt0 + value1 * dt1) / timedif;
-      infoStreamPrint(LOG_EVENTS_V, 0, "delayImpl: Linear interpolation of %g between %g and %g", timeStamp, time0, time1);
-      infoStreamPrint(LOG_EVENTS_V, 0, "delayImpl: Linear interpolation of %g value: %g and %g = %g", timeStamp, value0, value1, retVal);
+    }
+    /* linear interpolation */
+    else
+    {
+      timedif = time1 - time0;
+      dt0 = time1 - timeStamp;
+      dt1 = timeStamp - time0;
+      retVal = (value0 * dt0 + value1 * dt1) / timedif;
       return retVal;
     }
   }
+}
 
+
+/**
+ * @brief Returns value of zero crossing at current simulation time.
+ *
+ * @param data            Storing all simulation/model data.
+ * @param threadData      Used for error handling.
+ * @param exprNumber      Index of delay.
+ * @param relationIndex   Index of relation used for zero crossing.
+ * @param delayValue      Value to store in delay ringbuffer.
+ * @param delayTime       Time to delay expValue.
+ * @param delayMax        Maximum allowed delay time, defaults to delayTime.
+ * @return double         Value of zeroCrossing at current simulation time.
+ */
+double delayZeroCrossing(DATA* data, threadData_t *threadData, unsigned int exprNumber, unsigned int relationIndex, double delayValue, double delayTime, double delayMax)
+{
+  int foundEvent;
+  int eventPos;
+  double zeroCrossingValue;
+  double time = data->localData[0]->timeValue;
+
+  RINGBUFFER* delayStruct = data->simulationInfo->delayStructure[exprNumber];
+  zeroCrossingValue = data->simulationInfo->zeroCrossingsPre[relationIndex];
+
+  if (ringBufferLength(delayStruct) == 0)
+  {
+    return zeroCrossingValue;
+  }
+
+  /* Find first event on ring buffer */
+  findTime(time - delayTime, delayStruct, &foundEvent);
+
+  /* Flip sign of ZC if an event was found */
```

Comment:
why not just return 1 and -1? does this work if the zeroCrossingValue is 0? or is it always just 1 and -1?